### DISCUSSION TITLE
Collect commands that should be templates

### DISCUSSION THREAD

**Discussion by wsmelton** [dataplat/dbatools#7492] [Contributors]
I'm looking to start collecting commands or snippets that we want to now use as the standard in commands. Since we created the [templates](https://github.com/sqlcollaborative/dbatools-templates) repo many patterns we had started with have changed now.

The goal moving forward will be to drop these snippets and command templates into proper snippets for support with VS Code. This is a one started years ago that is now outdated as well, [dbatools.code-snippets](https://github.com/sqlcollaborative/dbatools/blob/development/.vscode/dbatools.code-snippets). Once we have this updated we can drop the template repo since it is no longer maintained.

An example. Those commands that directly support piping from the parent `Get` function could live on this pattern used in `Remove-DbaDbSequence`:

```powershell
    begin {
        $sequences = @( )
    }

    process {
        if ($SqlInstance) {
            $params = $PSBoundParameters
            $null = $params.Remove('WhatIf')
            $null = $params.Remove('Confirm')
            $sequences = Get-DbaDbSequence @params
        } else {
            $sequences += $InputObject
        }
    }
```

This gets turned into a snippet formatted block:

```json
"basic begin and process block for piping from Get function": {
  "prefix": "dbatPipeFromGet",
  "body": [
    "",
    "    begin {",
    "        \\$${1:object name}= @( )",
    "    }",
    "",
    "    process {",
    "        if (\\$SqlInstance) {",
    "            \\$params = \\$PSBoundParameters",
    "            \\$null = \\$params.Remove('WhatIf')",
    "            \\$null = \\$params.Remove('Confirm')",
    "            \\$${1} = Get-Dba${2:Function name} @params",
    "        } else {",
    "            \\$${1} += \\$InputObject",
    "        }",
    "    }"
  ],
  "description": "basic begin and process block for piping from Get function"
}
```

The `${x:description}` is annotations in snippets that will be tab accessible once it is inserted into the Editor of Code. When you repeat the use of the `${x}` it will auto-fill that value, only requires entering it once and done.

@potatoqualitee @niphlod @andreasjordan looking to replies added to this discussion with the link to the command (as a whole) or the snippet (permalink) of code that we want to use (or like 🧇 ).  Once we have a decent amount I'll create PRs to get them formatted and added into the snippet file.

Once I'm able to start working on Pester 5 conversion I'll add snippet support for that as well.

**Comment by andreasjordan** [comment]
Great idea! Can we have some kind of documentation to each snippet with some information about the "why"? Because sometimes I see code in a way that I would not use. After asking other contributors about the why and hearing the explaination - it's totally clear why we have the code this way.

**Comment by andreasjordan** [comment]
A basic output snippet for Remove- commands would be helpful as well:
https://github.com/sqlcollaborative/dbatools/blob/ae65fdc56f98bd779cd7271ed6bf43d6a1513cf3/functions/Remove-DbaDbSequence.ps1#L101-L126

Here is an example about the "why" - the comment in the first line. It is not the case in every situation, but a developer should be aware of that.

**Comment by andreasjordan** [comment]
I don't know if we already agreed on the "best" way to use parameters but some snippets for that would be great as well:
https://github.com/sqlcollaborative/dbatools/blob/ae65fdc56f98bd779cd7271ed6bf43d6a1513cf3/functions/Remove-DbaDbSequence.ps1#L67-L83

**Reply by wsmelton** [reply]
Parameter set name should not be used on SqlInstance because it should be required no matter what parameter set exists in the command.
